Set bss param max inactivity

ajayparida opened the pull request:
To set bss parameter maximum inactivity duration after wihch AP may throws away a STA.

ajayparida committed:
net: wifi_mgmt: Support to configure AP mode parameter

Support to set BSS parameter at compile and run time.

Added support to configure `max_inactivity` BSS parameter.
Station inactivity timeout is the period for which AP may keep a client
in associated state while there is no traffic from that particular client.
If a non-zero value is set, AP may choose to disassociate the
client after the timeout.

Signed-off-by: Ajay Parida <ajay.parida@nordicsemi.no>

kartben commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -84,6 +84,11 @@ enum net_request_wifi_cmd {
 	NET_REQUEST_WIFI_CMD_VERSION,
 	/** Set RTS threshold */
 	NET_REQUEST_WIFI_CMD_RTS_THRESHOLD,
+	/* Set AP max inactivity time.
```
```suggestion
	/** Set AP max inactivity time.
```

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -866,7 +876,19 @@ struct wifi_mgmt_ops {
+	 *
+	 * @return 0 if ok, < 0 if error
+	 */
+	int (*ap_sta_max_inactivity)(const struct device *dev,
```
Adding separate APIs for each param isn't scalable, we should have a single API e.g., `ap_sta_config` or `ap_config` and then add few param's, WDYT?

ajayparida replied:
Added single API to handle AP mode parameter .configuration

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1352,6 +1352,34 @@ static int cmd_wifi_ap_sta_disconnect(const struct shell *sh, size_t argc,
+
+	max_inactivity_time = (int)val;
+
+	if (net_mgmt(NET_REQUEST_WIFI_AP_STA_MAX_INACTIVITY, iface,
```
log the return value

ajayparida committed:
doc: release-notes: Configure BSS specific parameter

Added support to set BSS parameter.
Added support set BSS parameter maximum inactivity time.
Added support set BSS parameter inactivity poll feature.
Added support set BSS parameter maximum number of STA entries.

Signed-off-by: Ajay Parida <ajay.parida@nordicsemi.no>

rlubos commented on subsys/net/l2/wifi/Kconfig:
```diff
@@ -93,3 +93,12 @@ module-str = Log level for Wi-Fi Network manager module
 source "subsys/net/Kconfig.template.log_config.net"
 endif # WIFI_NM
+
+config WPA_SUPP_AP_STA_INACTIVITY_TIMEOUT
```
Was this added by mistake here? I don't think wpa supplicant-prefixed configs belong here, plus it doesn't seem to be used anywhere.

ajayparida committed:
net: wifi_mgmt: Support AP parameter configuration "Skip inactivity poll"

Support for configuration of AP parameter "Skip inactivity poll".
Only build time setting is supported.

Signed-off-by: Ajay Parida <ajay.parida@nordicsemi.no>

ajayparida committed:
net: wifi_mgmt: Support to configure AP parameter max_num_sta

Support to set BSS parameter "max_num_sta" at compile and run time

Added support to configure `max_num_sta` BSS parameter.
Maximum number of stations allowed in station table. New stations will be
rejected after the station table is full.

Signed-off-by: Ajay Parida <ajay.parida@nordicsemi.no>

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1352,6 +1355,49 @@ static int cmd_wifi_ap_sta_disconnect(const struct shell *sh, size_t argc,
+
+	if (!strncasecmp(argv[idx], "max_inactivity", 14)) {
+		while (idx < argc) {
+			if (!strncasecmp(argv[idx], "max_inactivity", 14)) {
```
Why the double comparison?

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1875,6 +1921,11 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_cmd_ap,
 		  cmd_wifi_ap_sta_disconnect,
 		  2, 0),
+	SHELL_CMD_ARG(config, NULL,
+		  "Configure AP parameter.\n"
```
`Configure AP, maximum STA inactivity timer`

ajayparida replied:
As there are multiple parameters to be configure, it's better to keep Configure AP parameter.

krish2718 commented on subsys/net/l2/wifi/Kconfig:
```diff
@@ -112,3 +112,12 @@ config WIFI_MGMT_AP_STA_SKIP_INACTIVITY_POLL
+config WIFI_MGMT_AP_MAX_NUM_STA
+	int "Maximum number of stations allowed in station table"
+	range 1 2007
+	default 128
```
For an Embedded RTOS, I would put a very lower default limit say 4?

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -356,6 +363,8 @@ struct wifi_connect_req_params {
 	/** Connect timeout in seconds, SYS_FOREVER_MS for no timeout */
 	int timeout;
+	/** Maximum inactivity time in seconds */
+	int32_t max_inactivity;
```
Why signed and not `uint32_t`? Would negative value has some special meaning?

ajayparida replied:
Type matching  has been kept to make it same as expected in wpa_supplicant. There is no other. 
In case of negative value being passed it will be reported as error.
Ïn supplicant side it's declared as "int ap_max_inactivity;

     /**
         * ap_max_inactivity - Timeout in seconds to detect STA's inactivity
         *
         * */
        int ap_max_inactivity;

krish2718 replied:
I don't see any reason for it to be `int` other than coming from the configuration file where there is only `INT` and no `UINT`.

and also `ap_max_inactivity` makes much more sense. And also why is it defined in both `wifi_connect_req_params` and `wifi_ap_config`?

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -693,6 +706,15 @@ struct wifi_channel_info {
+struct wifi_ap_config {
+	enum wifi_ap_param type;
+	/** Parameter used for setting ap_max_inactivity duration */
+	uint32_t ap_max_inactivity;
```
```suggestion
	uint32_t max_inactivity;
```

The comment to add `ap_` was for earlier generic structure, as the structure itself is for AP, no need for prefix.

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -693,6 +706,15 @@ struct wifi_channel_info {
+/** Wi-Fi AP configuration parameter*/
+struct wifi_ap_config {
+	enum wifi_ap_param type;
+	/** Parameter used for setting ap_max_inactivity duration */
```
```suggestion
	/** Parameter used for setting maximum inactivity duration for Stations */

krish2718 commented on subsys/net/l2/wifi/Kconfig:
```diff
@@ -93,3 +93,31 @@ module-str = Log level for Wi-Fi Network manager module
+
+config WIFI_MGMT_AP_MAX_NUM_STA
+	int "Maximum number of stations allowed in station table"
+	range 1 4
```
```suggestion
	range 1 2007
```

krish2718 replied:
I am fine with not having this change and limiting to 4, but it's easier to understand 2007 than 4.

krish2718 commented on subsys/net/l2/wifi/Kconfig:
```diff
@@ -93,3 +93,31 @@ module-str = Log level for Wi-Fi Network manager module
+	range 1 4
+	default 1
+	help
+	  Maximum number of stations allowed in station table. New stations will be
+	  rejected after the station table is full. IEEE 802.11 has a limit of 2007
+	  different association IDs, so this number should not be larger than that.
```
```suggestion
	  Maximum number of stations allowed in station table. New stations will be
	  rejected after the station table is full.
```

isn't the behaviour after reaching max dependent on the driver/NM's implementation?

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1352,6 +1352,59 @@ static int cmd_wifi_ap_sta_disconnect(const struct shell *sh, size_t argc,
+			param.type = WIFI_AP_CONFIG_PARAM_MAX_INACTIVITY;
+			param_valid = true;
+		}
+		if (!strncasecmp(argv[idx], "max_num_sta", 11)) {
```
```suggestion
		else if (!strncasecmp(argv[idx], "max_num_sta", 11)) {
```

krish2718 replied:
I missed that you are using a single shell command to configure all params, in which case please either use separate individual commands or move to `getopt` like we do for `connect`, `ap_enable` etc that can take multiple params.

krish2718 commented on include/zephyr/net/wifi.h:
```diff
@@ -27,6 +27,9 @@
 #define WIFI_LISTEN_INTERVAL_MAX 65535
 
+/** AP mode maximum number of STA entries */
+#define WIFI_AP_MODE_MAX_NUM_STA 4
```
This looks unused and enforcing a random limit, so, may be remove?

ajayparida replied:
Added the check for this value

krish2718 replied:
As discussed, we have decided to enforce 802.11 protocol limits, 4 is just the default. And we already have a Kconfig: `WIFI_MGMT_AP_MAX_NUM_STA`, so, please use that.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1875,6 +1939,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_cmd_ap,
 		  cmd_wifi_ap_sta_disconnect,
 		  2, 0),
+	SHELL_CMD_ARG(config, NULL,
+		  "Configure AP parameter.\n"
```
```suggestion
		  "Configure AP parameters.\n"
```

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1875,6 +1939,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_cmd_ap,
+	SHELL_CMD_ARG(config, NULL,
+		  "Configure AP parameter.\n"
+		  "-i --max_inactivity=<time duration in seconds>\n"
+		  "-s --max_num_sta=<maximum no of stations> (1-4)\n"
```
```suggestion
		  "-s --max_num_sta=<maximum number of stations>\n"
```

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1875,6 +1939,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_cmd_ap,
 		  2, 0),
+	SHELL_CMD_ARG(config, NULL,
+		  "Configure AP parameter.\n"
+		  "-i --max_inactivity=<time duration in seconds>\n"
```
```suggestion
		  "-i --max_inactivity=<time duration (in seconds)>\n"
```

krish2718 reviewed (commented):
Minor comments

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1352,6 +1353,79 @@ static int cmd_wifi_ap_sta_disconnect(const struct shell *sh, size_t argc,
+					    err);
+				return err;
+			}
+			params->max_num_sta = (uint32_t)val;
```
this should also have a boundary check `WIFI_MGMT_AP_MAX_NUM_STA`

krish2718 replied:
you can fail early, rather than relying on wifi_mgmt

krish2718 commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -435,6 +435,44 @@ static int wifi_ap_sta_disconnect(uint32_t mgmt_request, struct net_if *iface,
+		if (((int)params->max_num_sta > WIFI_AP_MODE_MAX_NUM_STA) ||
+		    ((int)params->max_num_sta < 0)) {
+			return -EINVAL;
+		}
```
This warrants log to increase the Kconfig

krish2718 commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -435,6 +435,44 @@ static int wifi_ap_sta_disconnect(uint32_t mgmt_request, struct net_if *iface,
+	}
+
+	if (params->type & WIFI_AP_CONFIG_PARAM_MAX_INACTIVITY) {
+		if ((int)params->max_inactivity < 0) {
```
when will this be true? The param itself is unsigned? And what is user gives very high unsigned value, this would be still be valid but due to `(int)` it now becomes invalid. This whole casting needs a rethink.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1352,6 +1352,94 @@ static int cmd_wifi_ap_sta_disconnect(const struct shell *sh, size_t argc,
+			params->type |= WIFI_AP_CONFIG_PARAM_MAX_NUM_STA;
+			break;
+		case 'h':
+			return SHELL_CMD_HELP_PRINTED;
```
This basically returns 1 and is typically used after printing `shell_help` see https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20SHELL_CMD_HELP_PRINTED&type=code 

and I see that the return value itself isn't used anywhere, just a convention to return non-zero value I guess.

jukkar commented on include/zephyr/net/wifi.h:
```diff
@@ -470,6 +470,14 @@ static inline const char *wifi_ps_get_config_err_code_str(int16_t err_no)
 	return "<unknown>";
 }
 
+/** Wi-Fi AP mode configuration parameter */
```
I just fixed lot of documentation issues in  #72992 and noticed that `@brief` should be added here to get better coverage.
So could you change this to `/** @brief Wi-Fi AP mode configuration parameter */`

ajayparida committed:
net: l2: wifi: Fix for typo

Typo: "RTS threashold" changed to "RTS threshold".

Signed-off-by: Ajay Parida <ajay.parida@nordicsemi.no>

ajayparida committed:
net: wifi: shell: Correction to connect command default case

For default case it should print the option character provided by user.
Extra shell_help removed as it's being called in the caller function.

Signed-off-by: Ajay Parida <ajay.parida@nordicsemi.no>

jukkar commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -435,6 +435,40 @@ static int wifi_ap_sta_disconnect(uint32_t mgmt_request, struct net_if *iface,
+
+	if (params->type & WIFI_AP_CONFIG_PARAM_MAX_NUM_STA) {
+		if (params->max_num_sta > CONFIG_WIFI_MGMT_AP_MAX_NUM_STA) {
+			LOG_INF("Maximum number of stations(%d) "
```
We should use `NET_INFO` here so that the output can be controlled by `CONFIG_NET_LOG` option. No need to change this now, this is just a FYI, I will send a PR changing all of these in networking code.

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -746,6 +753,13 @@ struct wifi_channel_info {
 
+/** @brief Wi-Fi AP configuration parameter */
+struct wifi_ap_config_params {
+	enum wifi_ap_config_param type;
```
This needs to be documented

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1363,6 +1363,74 @@ static int cmd_wifi_ap_sta_disconnect(const struct shell *sh, size_t argc,
 	return 0;
 }
 
+static int __wifi_ap_config_args_to_params(const struct shell *sh, size_t argc, char *argv[],
```
Avoid double underscores, they are not needed here.

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -52,6 +52,11 @@ extern "C" {
 #define WIFI_MGMT_BAND_STR_SIZE_MAX 8
 #define WIFI_MGMT_SCAN_MAX_BSS_CNT 65535
 
+#ifdef CONFIG_WIFI_MGMT_AP_STA_SKIP_INACTIVITY_POLL
+#define WIFI_MGMT_SKIP_INACTIVITY_POLL !!CONFIG_WIFI_MGMT_AP_STA_SKIP_INACTIVITY_POLL
+#else
+#define WIFI_MGMT_SKIP_INACTIVITY_POLL 0
+#endif /* CONFIG_WIFI_MGMT_AP_STA_SKIP_INACTIVITY_POLL */
```
I think we could simplify this to oneliner
```
#define WIFI_MGMT_SKIP_INACTIVITY_POLL IS_ENABLED(CONFIG_WIFI_MGMT_AP_STA_SKIP_INACTIVITY_POLL)
```

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1394,6 +1395,26 @@ static int __wifi_ap_config_args_to_params(const struct shell *sh, size_t argc,
 			params->type |= WIFI_AP_CONFIG_PARAM_MAX_INACTIVITY;
 			break;
+		case 's':
+			val = shell_strtol(optarg, 10, &err);
+			if (err) {
+				shell_error(sh, "Unable to parse input for max_num_sta (err %d)",
+					    err);
+				return err;
+			}
+			if (val < 0) {
+				PR_ERROR("Invalid value for max_num_sta %ld", val);
+				return -EINVAL;
+			}
```
We are duplicating this check multiple times, could we refactor this to a checker function that could be called from different case statements.

ajayparida committed:
net: wifi: shell: Support to print name

Support to print name of the argument along with the value.

Signed-off-by: Ajay Parida <ajay.parida@nordicsemi.no>
